include/seahorn/Expr/ExprOpFiniteMap.hh

```diff
@@ -0,0 +1,205 @@
+NOP(SET, "set", FUNCTIONAL, FiniteMapOp)
+
+} // namespace op
+
+namespace op {
+namespace finite_map {
+
+// --------------- finite map primitives ------------------------------------------
+inline Expr constFiniteMapValues(ExprVector &values) {
+  return mknary<CONST_FINITE_MAP_VALUES>(values.begin(), values.end());
+}
+
+inline Expr constFiniteMapKeys(ExprVector &keys) {
+  assert(keys.size() > 0);
+  return mknary<CONST_FINITE_MAP_KEYS>(keys.begin(), keys.end());
+}
+inline Expr constFiniteMap(ExprVector &keys, Expr vTy) {
+  return mk<CONST_FINITE_MAP>(constFiniteMapKeys(keys),
+                              mk<FINITE_MAP_VAL_DEFAULT>(vTy));
+}
+// construct when ALL the values of the map are known (they can be
+// variables)
+inline Expr constFiniteMap(ExprVector &keys, ExprVector &values) {
+  assert(keys.size() == values.size());
+  return mk<CONST_FINITE_MAP>(constFiniteMapKeys(keys),
+                              constFiniteMapValues(values));
+}
+
+inline Expr get(Expr map, Expr idx) { return mk<GET>(map, idx); }
+inline Expr set(Expr map, Expr idx, Expr v) { return mk<SET>(map, idx, v); }
+
+
+
+// --------------- finite map sort ------------------------------------------
+inline Expr valTy(Expr fmTy) { return fmTy->left(); }
+inline Expr keys(Expr fmTy) { return fmTy->right(); }
+
+
+
+// --------------- transformation to lambda functions ------------------------
+// \brief fresh map with unitialized values of type ty, returns '0'
+// TODO: change 0 by the same as unitialized memory?
+inline Expr mkEmptyMap(Expr ty, ExprFactory &efac) {
+  if(isOpX<INT_TY>(ty))
+    return mkTerm<mpz_class>(0, efac); // is this an int already?
+  else
+    return bv::bvnum(mkTerm<mpz_class>(0, efac), ty);
+}
+
+// creates a set of keys as a lambda function
+inline Expr mkKeys(ExprVector &keys, ExprFactory &efac) {
+
+  Expr x = bind::intConst(mkTerm<std::string>("x", efac));
+  // TODO: what do we use as variable name for the lambda function?
+
+  Expr lmd_bot = bind::abs<LAMBDA>(std::array<Expr, 1>{x},
+                                   mkTerm<mpz_class>(0, efac));
+  // up to here, it will be the same for all keysets
+
+  int count = 1;
+
+  Expr lmd_tmp = lmd_bot;
+
+  for (auto key : keys) {
+    Expr nA = mkTerm<mpz_class>(count, efac);
+    Expr cmp = mk<EQ>(key, x);
+    Expr ite = boolop::lite(cmp, nA, op::bind::betaReduce(lmd_tmp, x));
+    lmd_tmp = bind::abs<LAMBDA>(std::array<Expr, 1>{x}, ite);
+    count++;
+  }
+
+  return lmd_tmp;
+}
+
+// creates a map for keys and values, assuming that they are sorted
+// old make_map_batch_values
+inline Expr mkInitializedMap(ExprVector &keys, Expr vTy, ExprVector &values,
+                             ExprFactory &efac, Expr &lambda_keys) {
+
+  // assuming that there is a value for every key. If this is not available,
+  // "initialize" it with the default value for uninitialized memory
+  assert(keys.size() == values.size());
+
+  Expr x = bind::intConst(mkTerm<std::string>("x", efac));
```

Comment:
Lines 112-126 seem the same than `mkKeys`

Reply:
Fixed, thanks